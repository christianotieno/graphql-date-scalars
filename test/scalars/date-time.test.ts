import { Kind } from 'graphql';

import { stringify } from 'jest-matcher-utils';

import DateTimeScalar from '../../src/scalars/date-time';

describe('DateTime scalar', () => {
  const invalidDates = [
    // General
    'Invalid date',
    // Datetime with hours
    '2016-02-01T00Z',
    // Datetime with hours and minutes
    '2016-02-01T00:00Z',
    // Datetime with hours, minutes and seconds
    '2016-02-01T000059Z',
    // Datetime with hours, minutes, seconds and fractional seconds
    '2016-02-01T00:00:00.Z',
    // Datetime with hours, minutes, seconds, fractional seconds and timezone.
    '2015-02-24T00:00:00.000+0100'
  ];

  const validDates: [string, Date][] = [
    // Datetime with hours, minutes and seconds
    ['2016-02-01T00:00:15Z', new Date(Date.UTC(2016, 1, 1, 0, 0, 15))],
    ['2016-02-01T00:00:00-11:00', new Date(Date.UTC(2016, 1, 1, 11))],
    ['2017-01-07T11:25:00+01:00', new Date(Date.UTC(2017, 0, 7, 10, 25))],
    ['2017-01-07T00:00:00+01:20', new Date(Date.UTC(2017, 0, 6, 22, 40))],
    // Datetime with hours, minutes, seconds and fractional seconds
    ['2016-02-01T00:00:00.1Z', new Date(Date.UTC(2016, 1, 1, 0, 0, 0, 100))],
    ['2016-02-01T00:00:00.000Z', new Date(Date.UTC(2016, 1, 1, 0, 0, 0, 0))],
    ['2016-02-01T00:00:00.990Z', new Date(Date.UTC(2016, 1, 1, 0, 0, 0, 990))],
    ['2016-02-01T00:00:00.23498Z', new Date(Date.UTC(2016, 1, 1, 0, 0, 0, 234))],
    ['2017-01-07T11:25:00.450+01:00', new Date(Date.UTC(2017, 0, 7, 10, 25, 0, 450))]
  ];

  it('has a description', () => {
    expect(DateTimeScalar.description).toMatchSnapshot();
  });

  describe('serialization', () => {
    [{}, [], null, undefined, true].forEach(invalidInput => {
      it(`throws error when serializing ${stringify(invalidInput)}`, () => {
        expect(() => DateTimeScalar.serialize(invalidInput)).toThrowErrorMatchingSnapshot();
      });
    });

    [
      [new Date(Date.UTC(2016, 0, 1)), '2016-01-01T00:00:00.000Z'],
      [new Date(Date.UTC(2016, 0, 1, 14, 48, 10, 30)), '2016-01-01T14:48:10.030Z']
    ].forEach(([value, expected]) => {
      it(`serializes javascript Date ${stringify(value)} into ${stringify(expected)}`, () => {
        expect(DateTimeScalar.serialize(value)).toEqual(expected);
      });
    });

    it(`throws error when serializing invalid date`, () => {
      expect(() => DateTimeScalar.serialize(new Date('invalid date'))).toThrowErrorMatchingSnapshot();
    });
  });

  describe('value parsing', () => {
    validDates.forEach(([value, expected]) => {
      it(`parses date-string ${stringify(value)} into javascript Date ${stringify(expected)}`, () => {
        expect(DateTimeScalar.parseValue(value)).toEqual(expected);
      });
    });

    [4566, {}, [], true, null].forEach(invalidInput => {
      it(`throws an error when parsing ${stringify(invalidInput)}`, () => {
        expect(() => DateTimeScalar.parseValue(invalidInput)).toThrowErrorMatchingSnapshot();
      });
    });

    invalidDates.forEach(dateString => {
      it(`throws an error parsing an invalid date-string ${stringify(dateString)}`, () => {
        expect(() => DateTimeScalar.parseValue(dateString)).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('literal parsing', () => {
    validDates.forEach(([value, expected]) => {
      const literal = {
        kind: Kind.STRING,
        value
      };

      it(`parses literal ${stringify(literal)} into javascript Date ${stringify(expected)}`, () => {
        expect(DateTimeScalar.parseLiteral(literal, {})).toEqual(expected);
      });
    });

    invalidDates.forEach(value => {
      const invalidLiteral = {
        kind: Kind.STRING,
        value
      };

      it(`errors when parsing invalid literal ${stringify(invalidLiteral)}`, () => {
        expect(() => DateTimeScalar.parseLiteral(invalidLiteral, {})).toThrowErrorMatchingSnapshot();
      });
    });

    [
      {
        kind: Kind.FLOAT,
        value: '5'
      },
      {
        kind: Kind.BOOLEAN,
        value: false
      }
    ].forEach(literal => {
      it(`errors when parsing invalid literal ${stringify(literal)}`, () => {
        expect(() => DateTimeScalar.parseLiteral(literal, {})).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
