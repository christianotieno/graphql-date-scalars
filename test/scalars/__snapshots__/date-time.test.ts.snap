// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTime scalar has a description 1`] = `"A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar."`;

exports[`DateTime scalar literal parsing errors when parsing invalid literal {"kind": "BooleanValue", "value": false} 1`] = `"DateTime cannot represent non string type BooleanValue"`;

exports[`DateTime scalar literal parsing errors when parsing invalid literal {"kind": "FloatValue", "value": "5"} 1`] = `"DateTime cannot represent non string type FloatValue"`;

exports[`DateTime scalar literal parsing errors when parsing invalid literal {"kind": "StringValue", "value": "2015-02-24T00:00:00.000+0100"} 1`] = `"DateTime cannot represent an invalid date-time-string 2015-02-24T00:00:00.000+0100."`;

exports[`DateTime scalar literal parsing errors when parsing invalid literal {"kind": "StringValue", "value": "2016-02-01T00:00:00.Z"} 1`] = `"DateTime cannot represent an invalid date-time-string 2016-02-01T00:00:00.Z."`;

exports[`DateTime scalar literal parsing errors when parsing invalid literal {"kind": "StringValue", "value": "2016-02-01T00:00Z"} 1`] = `"DateTime cannot represent an invalid date-time-string 2016-02-01T00:00Z."`;

exports[`DateTime scalar literal parsing errors when parsing invalid literal {"kind": "StringValue", "value": "2016-02-01T000059Z"} 1`] = `"DateTime cannot represent an invalid date-time-string 2016-02-01T000059Z."`;

exports[`DateTime scalar literal parsing errors when parsing invalid literal {"kind": "StringValue", "value": "2016-02-01T00Z"} 1`] = `"DateTime cannot represent an invalid date-time-string 2016-02-01T00Z."`;

exports[`DateTime scalar literal parsing errors when parsing invalid literal {"kind": "StringValue", "value": "Invalid date"} 1`] = `"DateTime cannot represent an invalid date-time-string Invalid date."`;

exports[`DateTime scalar serialization throws error when serializing [] 1`] = `"DateTime cannot represent non-date type"`;

exports[`DateTime scalar serialization throws error when serializing {} 1`] = `"DateTime cannot represent non-date type"`;

exports[`DateTime scalar serialization throws error when serializing invalid date 1`] = `"DateTime cannot represent an invalid Date instance"`;

exports[`DateTime scalar serialization throws error when serializing null 1`] = `"DateTime cannot represent non-date type"`;

exports[`DateTime scalar serialization throws error when serializing true 1`] = `"DateTime cannot represent non-date type"`;

exports[`DateTime scalar serialization throws error when serializing undefined 1`] = `"DateTime cannot represent non-date type"`;

exports[`DateTime scalar value parsing throws an error parsing an invalid date-string "2015-02-24T00:00:00.000+0100" 1`] = `"DateTime cannot represent an invalid date-time-string 2015-02-24T00:00:00.000+0100."`;

exports[`DateTime scalar value parsing throws an error parsing an invalid date-string "2016-02-01T00:00:00.Z" 1`] = `"DateTime cannot represent an invalid date-time-string 2016-02-01T00:00:00.Z."`;

exports[`DateTime scalar value parsing throws an error parsing an invalid date-string "2016-02-01T00:00Z" 1`] = `"DateTime cannot represent an invalid date-time-string 2016-02-01T00:00Z."`;

exports[`DateTime scalar value parsing throws an error parsing an invalid date-string "2016-02-01T000059Z" 1`] = `"DateTime cannot represent an invalid date-time-string 2016-02-01T000059Z."`;

exports[`DateTime scalar value parsing throws an error parsing an invalid date-string "2016-02-01T00Z" 1`] = `"DateTime cannot represent an invalid date-time-string 2016-02-01T00Z."`;

exports[`DateTime scalar value parsing throws an error parsing an invalid date-string "Invalid date" 1`] = `"DateTime cannot represent an invalid date-time-string Invalid date."`;

exports[`DateTime scalar value parsing throws an error when parsing [] 1`] = `"DateTime cannot represent non string type []"`;

exports[`DateTime scalar value parsing throws an error when parsing {} 1`] = `"DateTime cannot represent non string type {}"`;

exports[`DateTime scalar value parsing throws an error when parsing 4566 1`] = `"DateTime cannot represent non string type 4566"`;

exports[`DateTime scalar value parsing throws an error when parsing null 1`] = `"DateTime cannot represent non string type null"`;

exports[`DateTime scalar value parsing throws an error when parsing true 1`] = `"DateTime cannot represent non string type true"`;
